### PR TITLE
Mark SafeHaskell explicitly

### PR DESCRIPTION
**Pull request by phadej**
This fixes a transient failure in compiling e.g. `http-link-header`
which relied on `network-uri` to be Safe/Trustworthy.

Also use `DeriveLift`, this fixes e.g. a clear error (though only
warning from GHC point-of-view), when GHC-8.10 is used:

    Network/URI.hs:1381:10: warning: [-Wmissing-methods]
        • No explicit implementation for
            ‘liftTyped’
        • In the instance declaration for ‘Lift URI’
         |
    1381 | instance Lift URI where
         |

There's an issue with GHC-8.10 release candidate, which I hope to fix for release proper: https://gitlab.haskell.org/ghc/ghc/issues/17813

### CODE REVIEW

File: network-uri.cabal [hunk 1]
```diff
@@ -67,7 +67,7 @@ library
   build-depends:
     base >= 3 && < 5,
     deepseq >= 1.1 && < 1.5,
-    parsec >= 3.0 && < 3.2
+    parsec >= 3.1.12.0 && < 3.2
```
Comment (ezrakilty) [inline network-uri.cabal]: Why this change? I see that version 3.1.12.0 is labelled `SafeHaskell: None` whereas parsec-3.1.3 is the first version I see that has `SafeHaskell: Inferred`, and parsec-3.1.13.0 seems to be the first version explicitly labelled as `SafeHaskell: Safe`.

I guess there's a ten-year gap between 3.0 and 3.1.12.0, so this could actually force people to upgrade.

File: Network/URI.hs [hunk 2]
```diff
@@ -1,9 +1,15 @@
...
+{-# LANGUAGE DeriveLift, StandaloneDeriving #-}
 #else
-{-# LANGUAGE TemplateHaskellQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+#endif
+#if MIN_VERSION_template_haskell(2,12,0) && MIN_VERSION_parsec(3,13,0)
+{-# LANGUAGE Safe #-}
+#elif __GLASGOW_HASKELL__ >= 702
+{-# LANGUAGE Trustworthy #-}
 #endif
```
Comment (ezrakilty) [inline Network/URI.hs]: If I understand this section correctly:

 * We used to use TemplateHaskellQuotes in GHC >= 8 because it was Safe-inferrable, but in GHC < 8 we forewent that nicety because it wasn't available.
 * Now we don't bother restricting from TemplateHaskell to TemplateHaskellQuotes, but instead we forcibly declare the module Trustworthy on GHC versions between 7.0.2 and 8.

It's a bit hard to think through all the CPP conditions so I want to make sure I have some grasp of it...

**Comment by ezrakilty** [comment]
Thanks! I think we can get this in, though I want to make sure I understand it first.

File: Network/URI.hs (hunk 2 above)
Comment (phadej) [reply Network/URI.hs]: - We don't need `TemplateHaskellQuotes` nor `TemplateHaskell` on GHC-8+ as `DeriveLift` does the needed job in this module
- `StandaloneDeriving` is to avoid bigger CPP mess in the module body. I.e. no need to try to squint `deriving (Lift)` into data type definition.
- We can be `Safe` when `template-haskell` (version bundled with GHC-8.2) and `parsec` (3.13.0.0 is the first one explicitly marked safe) are safe. That check should technically have `__GLASGOW_HASKELL__ >= 800` clause too, but err.
-  Otherwise we are just `Trustworthy`.
- Except on GHC-7.0 we aren't anything, as SafeHaskell is GHC-7.2+ thing.

File: network-uri.cabal (hunk 1 above)
Comment (phadej) [reply network-uri.cabal]: 3.1.13.0 doesn't support GHC-7.0 / 7.2.

And yes, there is little incentive to not use at least 3.1.12 or later, as the support goes very far back GHC-version wise, and there shouldn't be obstacles to upgrade. Version `3.1.0` is uploaded on by DerekElkins at Thu Mar 4 04:12:27 UTC 2010; almost a decade ago.

After that changes were only non-breaking additions, so the upgrade path should be smooth for anyone.

**Comment by phadej** [comment]
I also made revisions to `network-uri-2.6.2.0` Ihttps://hackage.haskell.org/package/network-uri-2.6.2.0/revisions/) and `network-uri-2.7.0.0` versions (https://hackage.haskell.org/package/network-uri-2.7.0.0/revisions/), as they are broken with GHC-8.10 (though compile, missing methods ought to be an error).

I think there should be tight upper bound on `template-haskell`, as its interface does change (even the usage in `network-uri` is quite small, still, here's an example that you cannot be ever too sure).

```
% cabal build -w ghc-8.10.1
Resolving dependencies...
Build profile: -w ghc-8.10.0.20200123 -O1
In order, the following will be built (use -v for more details):
 - network-uri-2.7.0.0 (lib) (first run)
Configuring library for network-uri-2.7.0.0..
Preprocessing library for network-uri-2.7.0.0..
Building library for network-uri-2.7.0.0..
[1 of 3] Compiling Network.URI      ( Network/URI.hs, /code/mess/nuri/network-uri-2.7.0.0/dist-newstyle/build/x86_64-linux/ghc-8.10.0.20200123/network-uri-2.7.0.0/build/Network/URI.o, /code/mess/nuri/network-uri-2.7.0.0/dist-newstyle/build/x86_64-linux/ghc-8.10.0.20200123/network-uri-2.7.0.0/build/Network/URI.dyn_o )

Network/URI.hs:1382:10: warning: [-Wmissing-methods]
    • No explicit implementation for
        ‘liftTyped’
    • In the instance declaration for ‘Lift URI’
     |
1382 | instance Lift URI where
     |          ^^^^^^^^

Network/URI.hs:1385:10: warning: [-Wmissing-methods]
    • No explicit implementation for
        ‘liftTyped’
    • In the instance declaration for ‘Lift URIAuth’
     |
1385 | instance Lift URIAuth where
     |     
```

**Comment by phadej** [comment]
Thanks for merging, can this be released? Without my builds of `github` fails on older GHCs (as some dependency incorrectly depended on `network-uri` to be safe)

**Comment by ezrakilty** [comment]
Yup, let's get this released. I'll try to get it done today.

**Comment by ezrakilty** [comment]
Voila: https://hackage.haskell.org/package/network-uri-2.6.3.0

**Comment by phadej** [comment]
@ezrakilty Thank you very much!
